File: app/src/main/java/com/commonpepper/photosen/ui/adapters/CategoriesAdapter.java
package com.commonpepper.photosen.ui.adapters;

import android.app.Activity;
import android.content.Intent;
import android.graphics.drawable.Drawable;
import android.os.Bundle;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.RelativeLayout;
import android.widget.TextView;

import com.commonpepper.photosen.R;
import com.commonpepper.photosen.ui.activities.SearchActivity;
import com.commonpepper.photosen.ui.activities.SinglePhotoActivity;
import com.google.android.flexbox.FlexboxLayout;
import com.google.android.flexbox.FlexboxLayoutManager;

import java.io.IOException;
import java.io.InputStream;

import androidx.annotation.NonNull;
import androidx.core.app.ActivityCompat;
import androidx.core.app.ActivityOptionsCompat;
import androidx.recyclerview.widget.RecyclerView;

public class CategoriesAdapter extends RecyclerView.Adapter<CategoriesAdapter.CategoriesViewHolder> {
    private static final String TAG = CategoriesAdapter.class.getSimpleName();

    private String[] categories;

    public CategoriesAdapter(String[] categories) {
        this.categories = categories;
    }

    @NonNull
    @Override
    public CategoriesViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.item_category, parent, false);
        CategoriesViewHolder viewHolder = new CategoriesViewHolder(view);
        return viewHolder;
    }

    @Override
    public void onBindViewHolder(@NonNull CategoriesViewHolder holder, int position) {
        holder.textView.setText(categories[position]);
        try {
            InputStream inputStream = holder.imageView.getContext().getAssets().open(categories[position] + ".jpg");
            holder.imageView.setImageDrawable(Drawable.createFromStream(inputStream, null));
        } catch (IOException e) {
            Log.e(TAG, e.toString());
        }
        holder.layout.setOnClickListener(v -> {
            Intent intent = new Intent(holder.textView.getContext(), SearchActivity.class);
            intent.putExtra(SearchActivity.TAG_SEARCHTAG, categories[position]);
            Bundle options = ActivityOptionsCompat.makeSceneTransitionAnimation((Activity)holder.textView.getContext(),
                    holder.textView, "sharedChip").toBundle();
            ActivityCompat.startActivity(holder.textView.getContext(), intent, options);
        });
    }

    @Override
    public int getItemCount() {
        return categories.length;
    }

    public static class CategoriesViewHolder extends RecyclerView.ViewHolder {
        public TextView textView;
        public ImageView imageView;
        public RelativeLayout layout;
        public CategoriesViewHolder(View view) {
            super(view);
            imageView = view.findViewById(R.id.category_image);
            textView = view.findViewById(R.id.category_text);
            layout = view.findViewById(R.id.category_layout);
        }
    }
}

File: app/src/main/java/com/commonpepper/photosen/ui/adapters/CategoriesAdapter.kt
package com.commonpepper.photosen.ui.adapters

import android.app.Activity
import android.content.Intent
import android.graphics.drawable.Drawable
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.RelativeLayout
import android.widget.TextView
import androidx.core.app.ActivityCompat
import androidx.core.app.ActivityOptionsCompat
import androidx.recyclerview.widget.RecyclerView.Adapter
import androidx.recyclerview.widget.RecyclerView.ViewHolder
import com.commonpepper.photosen.R.id
import com.commonpepper.photosen.R.layout
import com.commonpepper.photosen.ui.activities.SearchActivity
import kotlinx.android.synthetic.main.item_category.view.*
import java.io.IOException

class CategoriesAdapter(private val categories: Array<String>) : Adapter<CategoriesAdapter.CategoriesViewHolder>() {
    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): CategoriesViewHolder {
        val view: View = LayoutInflater.from(parent.context).inflate(layout.item_category, parent, false)
        return CategoriesViewHolder(view)
    }

    override fun onBindViewHolder(holder: CategoriesViewHolder, position: Int) {
        holder.itemView.categoryText.text = categories[position]
        try {
            val inputStream = holder.itemView.categoryImage.context.assets.open(categories[position] + ".jpg")
            holder.itemView.categoryImage.setImageDrawable(Drawable.createFromStream(inputStream, null))
        } catch (e: IOException) {
            Log.e(TAG, e.toString())
        }
        holder.itemView.categoryLayout.setOnClickListener {
            val intent = Intent(holder.itemView.categoryText.context, SearchActivity::class.java)
            intent.putExtra(SearchActivity.TAG_SEARCHTAG, categories[position])
            val options = ActivityOptionsCompat.makeSceneTransitionAnimation((holder.itemView.categoryText.context as Activity),
                    holder.itemView.categoryText, "sharedChip").toBundle()
            ActivityCompat.startActivity(holder.itemView.context, intent, options)
        }
    }

    override fun getItemCount(): Int {
        return categories.size
    }

    class CategoriesViewHolder(view: View) : ViewHolder(view)

    companion object {
        private val TAG = CategoriesAdapter::class.java.simpleName
    }

}